ignite/metrics/ssim.py

```diff
@@ -0,0 +1,171 @@
+            form expected by the metric.
+
+    Returns:
+        A Tensor with SSIM
+
+    Example:
+
+    To use with ``Engine`` and ``process_function``, simply attach the metric instance to the engine.
+    The output of the engine's ``process_function`` needs to be in the format of
+    ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return y_pred, y
+        engine = Engine(process_function)
+        metric = SSIM()
+        metric.attach(engine, "ssim")
+
+    If the output of the engine is not in the format above, ``output_transform`` argument can be used to transform it.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return {'prediction': y_pred, 'target': y, ...}
+
+        engine = Engine(process_function)
+
+        def output_transform(output):
+            # `output` variable is returned by above `process_function`
+            y_pred = output['prediction']
+            y = output['target']
+            return y_pred, y  # output format is according to `Accuracy` docs
+
+        metric = SSIM(output_transform=output_transform)
+        metric.attach(engine, "ssim")
+
+    The user even can use the metric with ``update`` and ``compute`` methods.
+
+    .. code-block:: python
+
+        >>> y_pred = torch.rand([16, 1, 16, 16])
+        >>> y = y_pred * 1.25
+        >>> ssim = SSIM()
+        >>> ssim.update((y_pred, y))
+        >>> ssim.compute()
+        tensor(0.9520)
+    """
+
+    def __init__(
+        self,
+        kernel_size: Sequence[int] = (11, 11),
+        sigma: Sequence[float] = (1.5, 1.5),
+        data_range: float = None,
+        k1: float = 0.01,
+        k2: float = 0.03,
+        output_transform: Callable = lambda x: x,
+    ):
+        if len(kernel_size) != 2 or len(sigma) != 2:
+            raise ValueError(
+                "Expected `kernel_size` and `sigma` to have the length of two."
+                f" Got kernel_size: {len(kernel_size)} and sigma: {len(sigma)}."
+            )
+
+        if any(x % 2 == 0 or x <= 0 for x in kernel_size):
+            raise ValueError(f"Expected `kernel_size` to have odd positive number. Got {kernel_size}.")
+
+        if any(y <= 0 for y in sigma):
+            raise ValueError(f"Expected `sigma` to have positive number. Got {sigma}.")
+
+        self.kernel_size = kernel_size
+        self.sigma = sigma
+        self.data_range = data_range
+        self.k1 = k1
+        self.k2 = k2
+        super(SSIM, self).__init__(output_transform=output_transform)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_batchwise_ssim = 0.0
+        self._num_examples = 0
+
+    def _gaussian_kernel(self, channel, kernel_size, sigma, device):
+        def gaussian(kernel_size, sigma, device):
```

Comment:
Maybe, it would be better to avoid function definition inside a function due to potential forking limitations...

Reply:
Okay, I will keep them separate.